File: week2/102_levelOrder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque

class Solution:
    def levelOrder1(self, root: TreeNode) -> List[List[int]]:
        res = []
        if not root:
            return res

        def helper(node, level):
            if len(res) == level:
                res.append([])

            res[level].append(node.val)

            if node.left:
                helper(node.left, level+1)
            if node.right:
                helper(node.right, level+1)

        helper(root, 0)
        return res

    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        res = []
        if not root:
            return res
        level = 0
        quene = deque([root,])
        while quene:
            res.append([])
            level_length = len(quene)

            for i in range(level_length):
                node = quene.popleft()
                res[level].append(node.val)

                if node.left:
                    quene.append(node.left)
                if node.right:
                    quene.append(node.right)
            level += 1
        return res

File: week2/102_levelOrder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>> res = new ArrayList<List<Integer>>();
    public List<List<Integer>> levelOrder(TreeNode root) {

        if (root == null) {
            return res;
        }
        helper(0, root);
        return res;

    }
    public void helper(int level, TreeNode node) {
        if (res.size() == level) {
            res.add(new ArrayList<Integer>());
        }
        res.get(level).add(node.val);

        if (node.left != null) {
            helper(node.left, level+1);
        }
        if (node.right != null) {
            helper(node.right, level+1);
        }

    }

    public List<List<Integer>> levelOrder_iteration(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if (root == null) {
            return res;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            List<Integer> temp = new ArrayList<>();
//            res.add(new ArrayList<Integer>());

            int level_length = queue.size();
            for (int i = 0; i < level_length; i++) {
                TreeNode node = queue.poll();
                temp.add(node.val);
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
            }
            res.add(temp);
        }
        return res;
    }
}